### PR TITLE
fix: reuse msgspec Encoder and Decoder instances

### PR DESCRIPTION
**Pull request by guacs**
[//]: # "By submitting this pull request, you agree to:"
[//]: # "- follow [Jolt's Code of Conduct](https://github.com/jolt-org/.github/blob/main/CODE_OF_CONDUCT.md)"
[//]: # "- follow [PSFs's Code of Conduct](https://www.python.org/psf/conduct/)"
[//]: # "- follow [Jolt's contribution guidelines](https://github.com/jolt-org/.github/blob/main/CONTRIBUTING.md)"

### Pull Request Checklist

- [ ] New code has 100% test coverage
- [ ] (If applicable) The prose documentation has been updated to reflect the changes introduced by this PR
- [ ] (If applicable) The reference documentation has been updated to reflect the changes introduced by this PR
- [ ] Pre-Commit Checks were ran and passed
- [ ] Tests were ran and passed

### Description
[//]: # "Please describe your pull request for new release changelog purposes"

- This PR reuses the `Encoder` and `Decoder` instances when using `msgspec` as the JSON serializer/deserializer.

### CODE REVIEW

**Comment by guacs** [comment]
As per the [msgspec docs](https://jcristharif.com/msgspec/perf-tips.html#reuse-encoders-decoders), it's a good idea to reuse the Encoder/Decoder for performance.

File: advanced_alchemy/config/engine.py [hunk 1]
```diff
@@ -15,23 +15,25 @@
...
 
 except ImportError:
     try:
-        from orjson import dumps as _encode_json  # type: ignore[no-redef]
+        from orjson import dumps as _encode_json
         from orjson import loads as decode_json  # type: ignore[no-redef]
 
         def encode_json(data: Any) -> str:
-            return _encode_json(data).decode("utf-8")
+            return _encode_json(data).decode("utf-8")  # type: ignore[no-any-return]
```
Comment (guacs) [inline advanced_alchemy/config/engine.py:32]: I have no idea why `mypy` kept complaining this is returning `Any`. `pyright` was able to resolve it properly, but `mypy` was not.

**Review by cofin — APPROVED** [review]

**Comment by cofin** [comment]
Thanks for the help finishing this up

**Comment by provinzkraut** [comment]
iirc @jcrist mentioned somewhere that the difference of doing this should be negligible if you're not specifying a type for the `Decoder`.

**Comment by jcrist** [comment]
This is correct. There used to be additional allocation-avoiding benefits, but these were removed in https://github.com/jcrist/msgspec/pull/402 to make the encoders/decoders threadsafe. Creating a `Decoder` beforehand is now only more efficient if you're passing in a type (to avoid re-parsing the type every call, which *can* be expensive for non-struct-types) or if passing in other configuration like `dec_hook` (although this is likely negligible for most users).

**Comment by guacs** [comment]
> This is correct. There used to be additional allocation-avoiding benefits, but these were removed in [jcrist/msgspec#402](https://github.com/jcrist/msgspec/pull/402) to make the encoders/decoders threadsafe. Creating a `Decoder` beforehand is now only more efficient if you're passing in a type (to avoid re-parsing the type every call, which _can_ be expensive for non-struct-types) or if passing in other configuration like `dec_hook` (although this is likely negligible for most users).

Is this the same case for `Encoder` as well?

**Comment by jcrist** [comment]
Yes, except that encoding is generally a faster process than decoding, so the constant time taken to process kwargs will be a higher fraction of the time per-call. So there might be more of a benefit for doing kwarg processing once when creating an `Encoder` rather than every call to `encode`.

In the case where you're passing no kwarg options, then there should be negligible benefit to creating an `Encoder` over just calling the top-level `encode` function.

**Comment by guacs** [comment]
> so the constant time taken to process kwargs will be a higher fraction of the time per-call

Oh, this is something I didn't consider at all. Basically, we could just prevent the call to `Encoder_init` right?

What about the memory allocation/deallocation though? Are they not expensive or are they negligible as well?

**Pull request merged**